src/mgr/ClusterState.cc

```diff
@@ -72,7 +72,17 @@ void ClusterState::ingest_pgstats(ref_t<MPGStats> stats)
   std::lock_guard l(lock);
 
   const int from = stats->get_orig_source().num();
-  pending_inc.update_stat(from, std::move(stats->osd_stat));
+  bool is_in = with_osdmap([from](const OSDMap& osdmap) {
+    return osdmap.is_in(from);
+  });
+
+  if (is_in) {
+    pending_inc.update_stat(from, std::move(stats->osd_stat));
+  } else {
+    struct osd_stat_t new_stat;
+    new_stat.seq = stats->osd_stat.seq;
+    pending_inc.update_stat(from, std::move(new_stat));  
```

Comment:
please check the implementation of `update_stat()`, i don't think the "empty" `new_stat` will end up with a no-op.

Reply:
@tchaikov  thanks for your comments, but I might not get your point. Could you be more specific?

From my understanding, currently we used the "empty" `new_stat` which only reserved the seq id of this OSD. Then `PGMap::stat_osd_add` would call `osd_sum.add(s)` but nothing would be updated and its seq id would be saved in `osd_last_seq`.

Reply:
in `osd_stat_t::add()`, we have
```c++
op_queue_age_hist.add(o.op_queue_age_hist);
```
in which the "0" bin in `op_queue_age_hist` is incremented if the new "v" is 0.

Reply:
@tchaikov  I just checked the code and did some testing. If `new_stat` is empty, its `op_queue_age_hist.h` which is a vector is also empty. So when calling `op_queue_age_hist.add(o.op_queue_age_hist);`, original  `op_queue_age_hist.h` won't be updated.

Also see the logs below:
```
2020-07-29 11:49:51.106100 7efe613ab700  1 stat_osd_add osd 0 stat osd_stat(0B used, 0B avail, 0B total, peers [] op hist [])
```

Reply:
i don't think you answered my concern. also, i cannot reach any conclusion by reading this logging message, as i have no idea where it is printed.

Reply:
Sorry, let me be more specific. 

From my understanding, your concern is that athough `osd_stat_t` is empty, it may also change something in `osd_stat_t::add()`, right?

But if `osd_stat_t` is empty, its `op_queue_age_hist.h` is also empty. `op_queue_age_hist.add(o.op_queue_age_hist);` won't increment any bin in `op_queue_age_hist` because the histogram vector `h` in `o.op_queue_age_hist` is empty. So does other stuff in `osd_stat_t::add()`. 

I added above log in `void PGMap::stat_osd_add(int osd, const osd_stat_t &s)` to print osd id and `osd_stat_t` s.

Reply:
please check `pow2_hist_t::add()`, if `v` is 0, `h[bin]++` will increment `h[0]` by 1.

Reply:
Ha, I see : ). There are two add() methods, one is `void add(int32_t v)`, the other is `void add(const pow2_hist_t& o)`. `osd_stat_t::add()` called the second one.